File: skills/core/scripts/graph-walk.py
#!/usr/bin/env python3
"""Tier 2 edge-graph walk for CORE memory retrieval, per DC-68/retrieval.md.

Given a seed unit, walks typed edges up to a hop cap, applying the R·S proxy
from priority.py for branch pruning. Deterministic alternative to LLM-by-hand
edge traversal — per DC-77, graph traversal logic ships in the plugin.

Library usage:
    from graph_walk import walk
    candidates = walk('_memories/dc-67-no-mcp.md',
                      memories_dir='_memories',
                      hops=2,
                      session_topics=['memory-architecture'])

CLI:
    python3 graph-walk.py <seed-unit-path> [--memories <dir>] [--hops 2]
                          [--budget 15] [--intent t1,t2] [--prune 0.3]
                          [--format json|text]
"""

from __future__ import annotations

import argparse
import json
import os
import sys
from dataclasses import dataclass, field
from datetime import date
from pathlib import Path
from typing import Iterator

# Import scoring from the co-located priority.py
_HERE = Path(__file__).parent
sys.path.insert(0, str(_HERE))
from priority import (  # noqa: E402
    Unit, load_unit, score_proxy_RS, extract_edges, parse_iso_date,
    SCORE_PRUNE_THRESHOLD,
)


@dataclass
class WalkCandidate:
    unit_id: str
    path: Path
    hop: int
    rs_score: float
    via_edge_type: str
    via_source: str


def _resolve_target(target: str, memories_dir: Path) -> Path | None:
    """Resolve an edge target string to an absolute unit file path."""
    t = target.strip()
    # Try as a direct path first
    direct = Path(t)
    if direct.is_absolute() and direct.exists():
        return direct
    # Strip .md suffix for matching
    stem = t.removesuffix(".md")
    # Try memories_dir / <stem>.md
    candidate = memories_dir / f"{stem}.md"
    if candidate.exists():
        return candidate
    # Try memories_dir / <target> as-is
    candidate2 = memories_dir / t
    if candidate2.exists():
        return candidate2
    return None


def walk(seed_path: str | Path,
         memories_dir: str | Path = "_memories",
         hops: int = 2,
         budget: int = 15,
         session_topics: list[str] | None = None,
         prune_threshold: float = SCORE_PRUNE_THRESHOLD,
         today: date | None = None) -> list[WalkCandidate]:
    """Walk typed edges from seed_path up to `hops` hops.

    Returns candidates ordered by (hop asc, rs_score desc), excluding the seed.
    Branches are pruned when R·S proxy < prune_threshold (default 0.3 per DC-69).
    """
    if today is None:
        today = date.today()
    if session_topics is None:
        session_topics = []

    memories_dir = Path(memories_dir).resolve()
    seed = load_unit(seed_path)

    visited: set[Path] = {Path(seed.path).resolve()}
    queue: list[tuple[int, Path, str, str]] = []  # (hop, path, edge_type, source_id)

    for e in extract_edges(seed):
        target_path = _resolve_target(e["target"], memories_dir)
        if target_path and target_path.resolve() not in visited:
            queue.append((1, target_path, e["type"], seed.id))

    results: list[WalkCandidate] = []

    while queue and len(results) < budget:
        hop, path, edge_type, source_id = queue.pop(0)
        resolved = path.resolve()
        if resolved in visited:
            continue
        visited.add(resolved)

        try:
            unit = load_unit(path)
        except Exception:
            continue

        rs = score_proxy_RS(unit, today)
        if rs < prune_threshold:
            continue  # branch pruned

        results.append(WalkCandidate(
            unit_id=unit.id,
            path=path,
            hop=hop,
            rs_score=rs,
            via_edge_type=edge_type,
            via_source=source_id,
        ))

        if hop < hops:
            for e in extract_edges(unit):
                target_path = _resolve_target(e["target"], memories_dir)
                if target_path and target_path.resolve() not in visited:
                    queue.append((hop + 1, target_path, e["type"], unit.id))

    results.sort(key=lambda c: (c.hop, -c.rs_score))
    return results


def _cli(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        description="Walk CORE memory unit edges for Tier 2 retrieval.")
    parser.add_argument("seed", help="Path to the seed unit file")
    parser.add_argument("--memories", default=None,
                        help="Path to _memories/ dir (default: sibling to seed)")
    parser.add_argument("--hops", type=int, default=2,
                        help="Maximum hops (default: 2; max recommended: 3)")
    parser.add_argument("--budget", type=int, default=15,
                        help="Maximum candidates in result set (default: 15)")
    parser.add_argument("--intent", default="",
                        help="Comma-separated session intent topics (unused in walk "
                             "but passed through for future alignment integration)")
    parser.add_argument("--prune", type=float, default=SCORE_PRUNE_THRESHOLD,
                        help=f"R·S pruning threshold (default: {SCORE_PRUNE_THRESHOLD})")
    parser.add_argument("--today", default=None,
                        help="Override today's date as YYYY-MM-DD")
    parser.add_argument("--format", choices=["json", "text"], default="json",
                        help="Output format (default: json)")
    args = parser.parse_args(argv)

    seed_path = Path(args.seed).resolve()
    if not seed_path.exists():
        print(f"error: seed unit not found: {seed_path}", file=sys.stderr)
        return 2

    if args.memories:
        memories_dir = Path(args.memories).resolve()
    else:
        memories_dir = seed_path.parent.resolve()

    if not memories_dir.is_dir():
        print(f"error: memories dir not found: {memories_dir}", file=sys.stderr)
        return 2

    today = parse_iso_date(args.today) if args.today else date.today()
    intent = [t.strip() for t in args.intent.split(",") if t.strip()]

    candidates = walk(
        seed_path,
        memories_dir=memories_dir,
        hops=args.hops,
        budget=args.budget,
        session_topics=intent,
        prune_threshold=args.prune,
        today=today,
    )

    if args.format == "json":
        out = [
            {
                "unit_id": c.unit_id,
                "path": str(c.path),
                "hop": c.hop,
                "rs_score": round(c.rs_score, 4),
                "via_edge_type": c.via_edge_type,
                "via_source": c.via_source,
            }
            for c in candidates
        ]
        print(json.dumps(out, indent=2))
    else:
        print(f"Walk from: {seed_path.name}  hops={args.hops}  prune={args.prune}")
        print(f"Date: {today}  memories: {memories_dir}")
        print("-" * 72)
        for c in candidates:
            print(f"  hop={c.hop}  rs={c.rs_score:.3f}  [{c.via_edge_type}]"
                  f"  {c.unit_id}  (via {c.via_source})")
        print(f"\n{len(candidates)} candidates (budget={args.budget})")

    return 0


if __name__ == "__main__":
    sys.exit(_cli(sys.argv[1:]))

File: skills/core/scripts/graph-walk.mjs
/**
 * Tier 2 edge-graph walk for CORE memory retrieval, per DC-68/retrieval.md.
 *
 * Given a seed unit, walks typed edges up to a hop cap, applying the R·S proxy
 * from priority.mjs for branch pruning. Deterministic alternative to LLM-by-hand
 * edge traversal — per DC-77, graph traversal logic ships in the plugin.
 *
 * Per DC-80 the plugin ships Node.js (.mjs) only.
 *
 * Library usage:
 *   import { walk } from './graph-walk.mjs';
 *   const candidates = walk('_memories/dc-67-no-mcp.md',
 *                           { memoriesDir: '_memories', hops: 2,
 *                             sessionTopics: ['memory-architecture'] });
 *
 * CLI:
 *   node graph-walk.mjs <seed-unit-path> [--memories <dir>] [--hops 2]
 *                       [--budget 15] [--intent t1,t2] [--prune 0.3]
 *                       [--format json|text]
 */

import { existsSync, readdirSync } from 'node:fs';
import { resolve, join, dirname, basename } from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  loadUnit, scoreProxyRS, extractEdges, parseIsoDate,
  SCORE_PRUNE_THRESHOLD,
} from './priority.mjs';

function resolveTarget(target, memoriesDir) {
  const t = target.trim();
  const direct = resolve(t);
  if (existsSync(direct)) return direct;
  const stem = t.replace(/\.md$/, '');
  const c1 = join(memoriesDir, `${stem}.md`);
  if (existsSync(c1)) return c1;
  const c2 = join(memoriesDir, t);
  if (existsSync(c2)) return c2;
  return null;
}

// Build an inverse edge index: target-stem -> [{sourcePath, sourceId, edgeType}]
// so walk() can surface inbound neighbors per Codex probe Round 2 finding.
// Scans top-level *.md files in memoriesDir (skips _prefixed and INDEX*).
export function buildInverseEdgeIndex(memoriesDir) {
  const inverse = new Map();
  let files;
  try { files = readdirSync(memoriesDir); } catch { return inverse; }
  for (const fname of files) {
    if (!fname.endsWith('.md')) continue;
    if (fname.startsWith('_') || fname.startsWith('INDEX')) continue;
    const filePath = join(memoriesDir, fname);
    let unit;
    try { unit = loadUnit(filePath); } catch { continue; }
    for (const e of extractEdges(unit)) {
      const stem = String(e.target).trim().replace(/\.md$/, '');
      if (!inverse.has(stem)) inverse.set(stem, []);
      inverse.get(stem).push({
        sourcePath: filePath,
        sourceId: unit.id,
        edgeType: e.type,
      });
    }
  }
  return inverse;
}

function inverseLookup(inverse, unitId, unitPath) {
  // Match by id first, then by filename stem (so units with no `id:` frontmatter
  // or with id-vs-filename mismatch still resolve).
  const out = [];
  if (unitId && inverse.has(unitId)) out.push(...inverse.get(unitId));
  const stem = basename(String(unitPath)).replace(/\.md$/, '');
  if (stem && stem !== unitId && inverse.has(stem)) out.push(...inverse.get(stem));
  return out;
}

export function walk(seedPath, {
  memoriesDir = '_memories',
  hops = 2,
  budget = 15,
  sessionTopics = [],
  pruneThreshold = SCORE_PRUNE_THRESHOLD,
  today = null,
} = {}) {
  const n = new Date();
  const t = today || new Date(Date.UTC(n.getFullYear(), n.getMonth(), n.getDate()));
  const mDir = resolve(memoriesDir);
  const seed = loadUnit(seedPath);
  const seedResolved = resolve(String(seedPath));
  const inverse = buildInverseEdgeIndex(mDir);

  const visited = new Set([seedResolved]);
  const queue = []; // [hop, path, edgeType, sourceId, direction]

  for (const e of extractEdges(seed)) {
    const tp = resolveTarget(e.target, mDir);
    if (tp && !visited.has(resolve(tp))) {
      queue.push([1, tp, e.type, seed.id, 'outbound']);
    }
  }
  for (const src of inverseLookup(inverse, seed.id, seedResolved)) {
    if (!visited.has(resolve(src.sourcePath))) {
      queue.push([1, src.sourcePath, src.edgeType, seed.id, 'inbound']);
    }
  }

  const results = [];

  while (queue.length && results.length < budget) {
    const [hop, path, edgeType, sourceId, direction] = queue.shift();
    const resolved = resolve(path);
    if (visited.has(resolved)) continue;
    visited.add(resolved);

    let unit;
    try { unit = loadUnit(path); } catch { continue; }

    const rs = scoreProxyRS(unit, t);
    if (rs < pruneThreshold) continue;

    results.push({
      unit_id: unit.id,
      path,
      hop,
      rs_score: rs,
      via_edge_type: edgeType,
      via_source: sourceId,
      edge_direction: direction,
    });

    if (hop < hops) {
      for (const e of extractEdges(unit)) {
        const tp = resolveTarget(e.target, mDir);
        if (tp && !visited.has(resolve(tp))) {
          queue.push([hop + 1, tp, e.type, unit.id, 'outbound']);
        }
      }
      for (const src of inverseLookup(inverse, unit.id, resolved)) {
        if (!visited.has(resolve(src.sourcePath))) {
          queue.push([hop + 1, src.sourcePath, src.edgeType, unit.id, 'inbound']);
        }
      }
    }
  }

  results.sort((a, b) => a.hop - b.hop || b.rs_score - a.rs_score);
  return results;
}

export function main(argv) {
  let seedArg = null;
  let memoriesArg = null;
  let hops = 2;
  let budget = 15;
  let intentStr = '';
  let prune = SCORE_PRUNE_THRESHOLD;
  let todayArg = null;
  let format = 'json';

  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--memories') { memoriesArg = argv[++i]; }
    else if (a === '--hops') { hops = parseInt(argv[++i], 10); }
    else if (a === '--budget') { budget = parseInt(argv[++i], 10); }
    else if (a === '--intent') { intentStr = argv[++i]; }
    else if (a === '--prune') { prune = parseFloat(argv[++i]); }
    else if (a === '--today') { todayArg = argv[++i]; }
    else if (a === '--format') { format = argv[++i]; }
    else if (!a.startsWith('--')) { seedArg = a; }
  }

  if (!seedArg) { process.stderr.write('usage: node graph-walk.mjs <seed-unit-path> [options]\n'); return 2; }

  const seedPath = resolve(seedArg);
  if (!existsSync(seedPath)) { process.stderr.write(`error: seed unit not found: ${seedPath}\n`); return 2; }

  const memoriesDir = memoriesArg ? resolve(memoriesArg) : dirname(seedPath);
  const today = todayArg ? (parseIsoDate(todayArg) || new Date()) : null;
  const sessionTopics = intentStr ? intentStr.split(',').map(s => s.trim()).filter(Boolean) : [];

  const candidates = walk(seedPath, { memoriesDir, hops, budget, sessionTopics, pruneThreshold: prune, today });

  if (format === 'json') {
    const out = candidates.map(c => ({
      unit_id: c.unit_id,
      path: String(c.path),
      hop: c.hop,
      rs_score: Math.round(c.rs_score * 10000) / 10000,
      via_edge_type: c.via_edge_type,
      via_source: c.via_source,
      edge_direction: c.edge_direction,
    }));
    console.log(JSON.stringify(out, null, 2));
  } else {
    console.log(`Walk from: ${basename(seedPath)}  hops=${hops}  prune=${prune}`);
    console.log(`Date: ${(today || new Date()).toISOString().slice(0, 10)}  memories: ${memoriesDir}`);
    console.log('-'.repeat(72));
    for (const c of candidates) {
      console.log(`  hop=${c.hop}  rs=${c.rs_score.toFixed(3)}  [${c.via_edge_type}]  ${c.unit_id}  (via ${c.via_source})`);
    }
    console.log(`\n${candidates.length} candidates (budget=${budget})`);
  }
  return 0;
}

// CLI entry guard. Set CORE_DEBUG_CLI_ENTRY=1 to log both strings if invocation
// silently no-ops (path-normalization, symlinks, OneDrive virtualization, etc.).
const _cliEntryArgv1 = process.argv[1];
const _cliEntrySelf = fileURLToPath(import.meta.url);
if (process.env.CORE_DEBUG_CLI_ENTRY) {
  process.stderr.write(`[cli-entry] argv[1]=${JSON.stringify(_cliEntryArgv1)}\n[cli-entry] self  =${JSON.stringify(_cliEntrySelf)}\n[cli-entry] match=${_cliEntryArgv1 === _cliEntrySelf}\n`);
}
if (_cliEntryArgv1 === _cliEntrySelf) {
  process.exit(main(process.argv.slice(2)));
}
